File: Golang/05. Inheritance Relationship/practice-17.go
/**
 * Practice 17: Vehicle Rental System — Inheritance
 * Task: A vehicle rental company rents out Cars, Bikes, and Trucks.
 *       All vehicles share Brand, Model, and Year of Manufacture.
 *       Each type has different rental pricing and discount rules.
 *
 * Rules:
 *   - Car  : $50/day. 10% discount if the car is older than 5 years.
 *   - Bike : $15/day. 15% discount if rental period is more than 7 days.
 *   - Truck: $100/day + $100/day additional maintenance fee.
 *
 * How to run:
 *   go run practice-17.go
 *
 * Key Concepts:
 *   - Struct embedding (Go's approach to inheritance)
 *   - Shared fields via embedded struct
 *   - Type-specific methods
 */

package main

import "fmt"

/** Base struct for all rental vehicles. */
type Vehicle17 struct {
	Brand string
	Model string
	Year  int
}

func (v *Vehicle17) ShowInfo() {
	fmt.Println("  Brand :", v.Brand)
	fmt.Println("  Model :", v.Model)
	fmt.Println("  Year  :", v.Year)
}

/** Car: $50/day, 10% discount if older than 5 years. */
type Car17 struct {
	Vehicle17
}

func (c *Car17) CalculateRentalCost(days int) float64 {
	cost := 50.0 * float64(days)
	if 2026-c.Year > 5 {
		cost *= 0.9
	}
	return cost
}

/** Bike: $15/day, 15% discount if rental > 7 days. */
type Bike17 struct {
	Vehicle17
}

func (b *Bike17) CalculateRentalCost(days int) float64 {
	cost := 15.0 * float64(days)
	if days > 7 {
		cost *= 0.85
	}
	return cost
}

/** Truck: $100/day + $100/day additional maintenance fee. */
type Truck17 struct {
	Vehicle17
}

func (t *Truck17) CalculateRentalCost(days int) float64 {
	return (100.0 + 100.0) * float64(days)
}

func main() {
	car := &Car17{Vehicle17{"Toyota", "Corolla", 2019}}
	bike := &Bike17{Vehicle17{"Yamaha", "R15", 2023}}
	truck := &Truck17{Vehicle17{"Volvo", "FH16", 2022}}

	days := 10

	fmt.Println("=== Car Rental ===")
	car.ShowInfo()
	fmt.Printf("  Rental (%d days): $%.2f\n", days, car.CalculateRentalCost(days))
	fmt.Println()

	fmt.Println("=== Bike Rental ===")
	bike.ShowInfo()
	fmt.Printf("  Rental (%d days): $%.2f\n", days, bike.CalculateRentalCost(days))
	fmt.Println()

	fmt.Println("=== Truck Rental ===")
	truck.ShowInfo()
	fmt.Printf("  Rental (%d days): $%.2f\n", days, truck.CalculateRentalCost(days))
	fmt.Println()

	fmt.Println("=== Discount Scenarios ===")

	oldCar := &Car17{Vehicle17{"Honda", "Civic", 2018}}
	fmt.Println("Old Car (2018):")
	oldCar.ShowInfo()
	fmt.Printf("  Rental (%d days): $%.2f\n", days, oldCar.CalculateRentalCost(days))
	fmt.Printf("  (10%% discount applied — car is %d years old)\n", 2026-oldCar.Year)
	fmt.Println()

	shortBike := &Bike17{Vehicle17{"Honda", "CBR", 2024}}
	shortDays := 5
	fmt.Printf("Bike for %d days (no discount):\n", shortDays)
	shortBike.ShowInfo()
	fmt.Printf("  Rental (%d days): $%.2f\n", shortDays, shortBike.CalculateRentalCost(shortDays))
	fmt.Println()

	longBike := &Bike17{Vehicle17{"Suzuki", "Gixxer", 2023}}
	longDays := 10
	fmt.Printf("Bike for %d days (15%% discount):\n", longDays)
	longBike.ShowInfo()
	fmt.Printf("  Rental (%d days): $%.2f\n", longDays, longBike.CalculateRentalCost(longDays))
}

File: Java/05. Inheritance Relationship/Practice17.java
/**
 * Practice 17: Vehicle Rental System — Inheritance
 * Task: A vehicle rental company rents out Cars, Bikes, and Trucks.
 *       All vehicles share Brand, Model, and Year of Manufacture.
 *       Each type has different rental pricing and discount rules.
 *
 * Rules:
 *   - Car  : $50/day. 10% discount if the car is older than 5 years.
 *   - Bike : $15/day. 15% discount if rental period is more than 7 days.
 *   - Truck: $100/day + $100/day additional maintenance fee.
 *
 * How to compile and run:
 *   javac Practice17.java
 *   java Practice17
 *
 * Key Concepts:
 *   - Inheritance (IS-A relationship)
 *   - Base class and subclasses
 *   - Shared properties via parent class
 */

public class Practice17 {

    /** Base class for all rental vehicles. */
    static class Vehicle {
        protected String brand;
        protected String model;
        protected int year;

        public void showInfo() {
            System.out.println("  Brand : " + brand);
            System.out.println("  Model : " + model);
            System.out.println("  Year  : " + year);
        }
    }

    /** Car: $50/day, 10% discount if older than 5 years. */
    static class Car extends Vehicle {
        private static final double BASE_RATE = 50.0;

        public Car(String brand, String model, int year) {
            this.brand = brand;
            this.model = model;
            this.year = year;
        }

        public double calculateRentalCost(int days) {
            double cost = BASE_RATE * days;
            if (2026 - year > 5) {
                cost *= 0.9;
            }
            return cost;
        }
    }

    /** Bike: $15/day, 15% discount if rental > 7 days. */
    static class Bike extends Vehicle {
        private static final double BASE_RATE = 15.0;

        public Bike(String brand, String model, int year) {
            this.brand = brand;
            this.model = model;
            this.year = year;
        }

        public double calculateRentalCost(int days) {
            double cost = BASE_RATE * days;
            if (days > 7) {
                cost *= 0.85;
            }
            return cost;
        }
    }

    /** Truck: $100/day + $100/day additional maintenance fee. */
    static class Truck extends Vehicle {
        private static final double BASE_RATE = 100.0;
        private static final double ADDITIONAL_FEE = 100.0;

        public Truck(String brand, String model, int year) {
            this.brand = brand;
            this.model = model;
            this.year = year;
        }

        public double calculateRentalCost(int days) {
            return (BASE_RATE + ADDITIONAL_FEE) * days;
        }
    }

    public static void main(String[] args) {
        Car car = new Car("Toyota", "Corolla", 2019);
        Bike bike = new Bike("Yamaha", "R15", 2023);
        Truck truck = new Truck("Volvo", "FH16", 2022);

        int days = 10;

        System.out.println("=== Car Rental ===");
        car.showInfo();
        System.out.printf("  Rental (%d days): $%.2f%n", days, car.calculateRentalCost(days));
        System.out.println();

        System.out.println("=== Bike Rental ===");
        bike.showInfo();
        System.out.printf("  Rental (%d days): $%.2f%n", days, bike.calculateRentalCost(days));
        System.out.println();

        System.out.println("=== Truck Rental ===");
        truck.showInfo();
        System.out.printf("  Rental (%d days): $%.2f%n", days, truck.calculateRentalCost(days));
        System.out.println();

        System.out.println("=== Discount Scenarios ===");

        Car oldCar = new Car("Honda", "Civic", 2018);
        System.out.println("Old Car (2018):");
        oldCar.showInfo();
        System.out.printf("  Rental (%d days): $%.2f%n", days, oldCar.calculateRentalCost(days));
        System.out.printf("  (10%% discount applied — car is %d years old)%n", 2026 - oldCar.year);
        System.out.println();

        Bike shortBike = new Bike("Honda", "CBR", 2024);
        int shortDays = 5;
        System.out.printf("Bike for %d days (no discount):%n", shortDays);
        shortBike.showInfo();
        System.out.printf("  Rental (%d days): $%.2f%n", shortDays, shortBike.calculateRentalCost(shortDays));
        System.out.println();

        Bike longBike = new Bike("Suzuki", "Gixxer", 2023);
        int longDays = 10;
        System.out.printf("Bike for %d days (15%% discount):%n", longDays);
        longBike.showInfo();
        System.out.printf("  Rental (%d days): $%.2f%n", longDays, longBike.calculateRentalCost(longDays));
    }
}
